### PR TITLE
feat: switch to compounding from consolidation requests

### PR DESCRIPTION
**Pull request by ensi321**
### Spec changes
- A consolidation request where source == target will now be considered as a "switch to compounding" request. The related validator will have its withdrawal credential switched to compounding in `processConsolidationRequest`
- `processConsolidationRequest` will also switch the request's target validator to compounding instead of waiting until `processPendingConsolidations` in epoch processing.
- Compound switching no longer applied during `applyDeposit`.
- Related spec PR: ethereum/consensus-specs#3918

### Engineering changes
- Removing `EpochTransitionCache.newCompoundingValidators` added in #7043 as compound switching now happens in block processing instead of epoch processing. cc @twoeths

### CODE REVIEW

**Comment by ensi321** [comment]
Switching target branch to `nc/devnet-4`. Need to consolidate several PRs into the same branch in order to pass `v1.5.0-alpha.7` spec test

File: packages/state-transition/src/block/processConsolidationRequest.ts [hunk 1]
```diff
@@ -1,42 +1,51 @@
...
   // Verify that source != target, so a consolidation cannot be used as an exit.
   if (sourceIndex === targetIndex) {
     return;
   }
 
   const sourceValidator = state.validators.get(sourceIndex);
   const targetValidator = state.validators.getReadonly(targetIndex);
-  const sourceWithdrawalAddress = sourceValidator.withdrawalCredentials.subarray(12);
   const currentEpoch = state.epochCtx.epoch;
+  const sourceWithdrawalAddressStr = toHex(sourceValidator.withdrawalCredentials.subarray(12));
```
Comment (nflaig) [inline packages/state-transition/src/block/processConsolidationRequest.ts]: why do we need to convert to hex here?

File: packages/state-transition/src/block/processConsolidationRequest.ts [hunk 2]
```diff
@@ -70,4 +79,56 @@ export function processConsolidationRequest(
...
+
+/**
+ * Determine if we should set consolidation target validator to compounding credential
+ */
+function isValidSwitchToCompoundRequest(
+  state: CachedBeaconStateElectra,
+  consolidationRequest: electra.ConsolidationRequest
+): boolean {
+  const {sourcePubkey, targetPubkey, sourceAddress} = consolidationRequest;
+  const sourceIndex = state.epochCtx.getValidatorIndex(sourcePubkey);
```
Comment (nflaig) [inline packages/state-transition/src/block/processConsolidationRequest.ts:95]: we could save the index lookup if we compare the pubkeys first, not sure if it's worth though but would match more closely to the spec

Comment (ensi321) [reply packages/state-transition/src/block/processConsolidationRequest.ts:95]: > we could save the index lookup if we compare the pubkeys first, not sure if it's worth though but would match more closely to the spec

Let's do the optimization later. This part of spec is sub-optimized and will very likely to be cleaned up in alpha 9 or 10. We can optimize when the spec gets refactored.

File: packages/state-transition/src/block/processConsolidationRequest.ts [hunk 3]
```diff
@@ -70,4 +78,56 @@ export function processConsolidationRequest(
...
+    return false;
+  }
+
+  // Switch to compounding requires source and target be equal
+  if (sourceIndex !== targetIndex) {
+    return false;
+  }
+
+  const sourceValidator = state.validators.getReadonly(sourceIndex);
+  const addressStr = toHex(sourceValidator.withdrawalCredentials.subarray(12));
```
Comment (nflaig) [inline packages/state-transition/src/block/processConsolidationRequest.ts]: same here, does not seem like the hex conversion is required

Comment (ensi321) [reply packages/state-transition/src/block/processConsolidationRequest.ts]: thanks

**Review by nflaig — APPROVED** [review]

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.23.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.23.0) :tada:

**Pull request merged**